### PR TITLE
Allow Range#=== and Range#cover? on Range

### PR DESCRIPTION
**Pull request by utilum**
ruby/ruby@989e07c features switching `Range#===` to use internal `r_cover_p`
instead of rubyland `include?`. This breaks expected behavior of
`ActiveSupport::CoreExt::Range` documented since at least 8b67a02.

This patch adds overrides on `Range#cover?` and `Range#===` and places all
three in a single module, `CompareWithRange`.
See failures at:
https://travis-ci.org/rails/rails/jobs/380939901#L1224-L1247

### CODE REVIEW

**Comment by matthewd** [comment]
If we go with this (see parallel discussion in #32945), I think we should also address `cover?`, and keep all three in the same file.

**Comment by utilum** [comment]
Thank you both.

I've added `cover?` and moved all three into a single file, in module `RangeCompareWithRange`.

Better?

**Comment by utilum** [comment]
PS: What do you think about moving `overlaps?` in with the others?

File: guides/source/active_support_core_extensions.md [hunk 1]
```diff
@@ -2889,9 +2889,9 @@ As the example depicts, the `:db` format generates a `BETWEEN` SQL clause. That
 
 NOTE: Defined in `active_support/core_ext/range/conversions.rb`.
 
-### `include?`
+### `===`, `include?`, and `range?`
```
Comment (eugeneius) [inline guides/source/active_support_core_extensions.md]: `range?` -> `cover?`

File: activesupport/lib/active_support/core_ext/range/include_range.rb [hunk 2]
```diff
@@ -1,25 +0,0 @@
-# frozen_string_literal: true
```
Comment (eugeneius) [inline activesupport/lib/active_support/core_ext/range/include_range.rb]: The existence of this file is documented, and users may depend on it:

http://guides.rubyonrails.org/v5.2.0/active_support_core_extensions.html#include-questionmark

If we want to remove it, it needs a deprecation cycle first. Here's an example of how that might look:

https://github.com/rails/rails/blob/64c88fb5d2caf3c34742a07394ac68b8377c4936/activesupport/lib/active_support/core_ext/object/to_json.rb

File: guides/source/active_support_core_extensions.md (hunk 1 above)
Comment (utilum) [reply guides/source/active_support_core_extensions.md]: Yes! of course.

File: activesupport/lib/active_support/core_ext/range/include_range.rb (hunk 2 above)
Comment (utilum) [reply activesupport/lib/active_support/core_ext/range/include_range.rb]: Thank you. Done.

File: activesupport/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,17 @@
+*   Allow Range#=== and Range#cover? on Range
+
+    ruby/ruby@989e07c features switching `Range#===` to use internal
+    `r_cover_p` instead of rubyland `include?`. This breaks expected behavior
+    of `ActiveSupport::CoreExt::Range` documented since at least 8b67a02.
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: This is excellent detail for anyone interested in why the change was necessary, but the changelog should be a high level description of what changed from the user's perspective - it's probably sufficient to say that `Range#cover?` can now be given a range like `Range#===` and `Range#include?`, and `Range#===` works correctly when given a range on Ruby 2.6.

File: activesupport/lib/active_support/core_ext/range/include_range.rb [hunk 4]
```diff
@@ -1,25 +1,7 @@
...
-        # 1...10 includes 1..9 but it does not include 1..10.
-        operator = exclude_end? && !value.exclude_end? ? :< : :<=
-        super(value.first) && value.last.send(operator, last)
-      else
-        super
-      end
-    end
-  end
-end
+ActiveSupport::Deprecation.warn "You have required `active_support/core_ext/range/include_range`. " \
```
Comment (eugeneius) [inline activesupport/lib/active_support/core_ext/range/include_range.rb]: [Every](https://github.com/rails/rails/blob/6c574ac58993512f975ddaf1f679c5956cc576df/activesupport/lib/active_support/core_ext/hash/compact.rb#L3) [other](https://github.com/rails/rails/blob/6c574ac58993512f975ddaf1f679c5956cc576df/activesupport/lib/active_support/core_ext/hash/transform_values.rb#L3) [file](https://github.com/rails/rails/blob/6c574ac58993512f975ddaf1f679c5956cc576df/activesupport/lib/active_support/core_ext/numeric/inquiry.rb#L3) in Active Support that emits a deprecation requires `active_support/deprecation` first; we should do the same here.

File: guides/source/active_support_core_extensions.md [hunk 5]
```diff
@@ -2889,9 +2889,9 @@ As the example depicts, the `:db` format generates a `BETWEEN` SQL clause. That
 
 NOTE: Defined in `active_support/core_ext/range/conversions.rb`.
 
-### `include?`
+### `===`, `include?`, and `cover?`
 
-The methods `Range#include?` and `Range#===` say whether some value falls between the ends of a given instance:
+The methods `Range#===`, `Range#include?` and `Range#cover?` say whether some value falls between the ends of a given instance:
```
Comment (eugeneius) [inline guides/source/active_support_core_extensions.md]: http://guides.rubyonrails.org/v5.2.0/api_documentation_guidelines.html#oxford-comma

> Please use the Oxford comma ("red, white, and blue", instead of "red, white and blue").

🤓

File: activesupport/lib/active_support/core_ext/range/include_range.rb (hunk 4 above)
Comment (utilum) [reply activesupport/lib/active_support/core_ext/range/include_range.rb]: Corrected.

File: activesupport/CHANGELOG.md (hunk 3 above)
Comment (utilum) [reply activesupport/CHANGELOG.md]: Cleaned up. I hope I did not go too far by also removing the module name.

File: guides/source/active_support_core_extensions.md (hunk 5 above)
Comment (utilum) [reply guides/source/active_support_core_extensions.md]: Oxford in a comma, it's serious ;)

File: activesupport/CHANGELOG.md (hunk 3 above)
Comment (eugeneius) [reply activesupport/CHANGELOG.md]: I think that's the right call - users don't interact with the module directly, they just call the methods that it modifies.

File: activesupport/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,15 @@
+*   Allow Range#=== and Range#cover? on Range
+
+    `Range#cover?` can now accept a range argument like `Range#include` and
+    `Range===`. `Range#===` works correctly on Ruby 2.6. `Range#include?` is moved
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: There's a missing `#` in `Range===` here.

File: activesupport/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,15 @@
...
+    into a new file, with these two methods.
+
+    To use: `require "active_support/core_ext/compare_range"`.
+
+    *Requiring core_ext/range/include_range now causes a deprecation warnning*
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: "warnning" -> "warning"

Can you also add `active_support/` to the start of the filename, to match the example above?

Comment (eugeneius) [reply activesupport/CHANGELOG.md]: We're telling the user that requiring this file is deprecated, but not what they need to do instead - it's pretty easy to _infer_ that they should now require `active_support/core_ext/compare_range`, but ideally we'd explicitly state that.

**Comment by eugeneius** [comment]
> What do you think about moving `overlaps?` in with the others?

I'd recommend leaving it where it is, mostly just to avoid inflicting the mild inconvenience of a deprecation warning on anyone cherry-picking that file on its own.

File: activesupport/CHANGELOG.md (hunk 7 above)
Comment (utilum) [reply activesupport/CHANGELOG.md]: Right on.
Sorry about all the typos.

File: activesupport/CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,14 @@
+*   Allow Range#=== and Range#cover? on Range
+
+    `Range#cover?` can now accept a range argument like `Range#include` and
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: `Range#include` -> `Range#include?`

File: activesupport/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,14 @@
...
+    `Range#===`. `Range#===` works correctly on Ruby 2.6. `Range#include?` is moved
+    into a new file, with these two methods.
+
+    *Requiring active_support_core_ext/range/include_range now causes a deprecation warning*
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: `active_support_core_ext` -> `active_support/core_ext`

File: activesupport/CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,14 @@
...
+    *Use `require "active_support/core_ext/compare_range"` instead.*
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]: `active_support/core_ext/compare_range` -> `active_support/core_ext/range/compare_range`

File: activesupport/CHANGELOG.md (hunk 9 above)
Comment (eugeneius) [reply activesupport/CHANGELOG.md]: "now causes a deprecation warning" -> "is now deprecated"

Comment (utilum) [reply activesupport/CHANGELOG.md]: Fixed.

File: activesupport/CHANGELOG.md (hunk 10 above)
Comment (utilum) [reply activesupport/CHANGELOG.md]: Fixed.

File: activesupport/CHANGELOG.md (hunk 8 above)
Comment (utilum) [reply activesupport/CHANGELOG.md]: Fixed.

**Review by eugeneius — APPROVED** [review]
